File: codingtests/algorithms/reverse_linked_list.py
class SLLNode:
    """ Definition for a singly-linked list node. """
    def __init__(self, val=0, next=None):
        self.val: int = val
        self.next: SLLNode = next


def print_linked_list(linked_list):
    nodes = []
    current = linked_list
    while current:
        nodes.append(str(current.val))
        current = current.next
    output = (" -> ".join(nodes) if nodes else "Empty List")
    print (output)
    return output


def reverse_list(input_linked_list):
    """
    Reverse a singly linked list iteratively.

    Time Complexity: O(n) - Iterate through list of length n once.
    Space Complexity: O(1) - Constant space utilized for pointers.
    """
    prev = None
    current = input_linked_list

    while current:
        next_node = current.next  # Temp storage of next node
        current.next = prev       # Reverse current node's pointer
        prev = current            # Move prev one forward
        current = next_node       # Move current one forward

    return prev  # new head of reversed list


def test_reverse_list():
    test_data_list = [
        {
            'input': SLLNode(1, SLLNode(2, SLLNode(3, SLLNode(4)))),
            'expected': '4 -> 3 -> 2 -> 1'
        },
        {'input': SLLNode(1), 'expected': '1'},
        {'input': SLLNode(), 'expected': '0'},
        {'input': [], 'expected': 'Empty List'},
    ]
    for count, test_data in enumerate(test_data_list):
        reversed_list = reverse_list(test_data['input'])
        actual = print_linked_list(reversed_list)
        assert actual == test_data['expected'], \
            f"Test #{count+1}: Expected {test_data['expected']}, actual: {actual}"
    
    print(f"PASSED {len(test_data_list)} TESTS")


if __name__ == "__main__":
    test_reverse_list()  # run tests

File: codingtests/algorithms/reverse_linked_list.js
/** Definition for a singly-linked list node. */
class SLLNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

/**
 * Reverses singly linked list (iteratively).
 * Time Complexity: O(n) - through list of length n exactly once.
 * Space Complexity: O(1) - Constant space for pointers.
 */
function reverseList(singlyLinkedList) {
    let prev = null;
    let current = singlyLinkedList;

    while (current !== null) {
        let nextNode = current.next; // Temporarily store the next node
        current.next = prev;         // Reverse the current node's pointer
        prev = current;              // Move prev one step forward
        current = nextNode;          // Move current one step forward
    }
    return prev; // prev new head of reversed list
}

function printList(singlyLinkedList) {
    const nodeList = [];
    let current = singlyLinkedList;
    while (current !== null) {
        nodeList.push(current.val);
        current = current.next;
    }
    console.log(nodeList.length > 0 ? nodeList.join(" -> ") : "Empty List");
}

// Usage
const inputSLList = new SLLNode(1, new SLLNode(2, new SLLNode(3, new SLLNode(4, new SLLNode(5)))));
console.log("Original List:");
printList(inputSLList);

const reversedSLL = reverseList(inputSLList);
console.log("Reversed List:");
printList(reversedSLL);
